# DS-GA 1008: Deep Learning, Spring 2020

# Homework Assignment 2

>*The search for truth is more precious than its possession.*
>
>Albert Einstein (1879 - 1955)

## 1. Fundamentals

### 1.1. Convolution

**Table 1** depicts two matrices:

> **Table 1**: Image Matrix (5 × 5) and a convolution kernel (3 × 3).
> 
> $$A=\left[\begin{matrix} 4 && 5 && 2 && 2 && 1\\
3 && 3 && 2 && 2 && 4\\
4 && 3 && 4 && 1 && 1\\
5 && 1 && 4 && 1 && 2\\
5 && 1 && 3 && 1 && 4\\
\end{matrix}\right] \ \ 
B=\left[\begin{matrix} 4 && 3 && 3\\
5 && 5 && 5\\
2 && 4 && 3\\
\end{matrix}\right]$$
>
>
>* The one on the left represents an 5 × 5 single-channel image A.
>
>
>* The one on the right represents a 3 × 3 convolution kernel B.


* **(a)** What is the **dimensionality of the output** if we forward propagate the image over the given convolution kernel with **no padding** and stride of 1?

#### **Answer:**

|Output dim|
|--|
|$$3 \times 3$$|

* **(b)** Give a general formula of the **output width** $O$ in terms of the:
  * input width $I$,
  
  * kernel width $K$,
  
  * stride $S$,
  
  * and padding $P$ (both in the beginning and in the end).
  
  Note that the same formula holds for the height.
  
  Make sure that your answer in part **(a)** is consistent with your formula.

#### **Answer:**

|Output width $O$|
|--|
|$$I - S(K-1)+ 2P$$|

Values from part **(a)**:

> $I=5$
>
> $K=3$
>
> $S=1$
>
> $P=0$

Replacing on the formula:
$$5 - 1(3-1)+ 2(0) = 3$$

Which is the same answer given in **(a)**

* **(c)** Compute the **output** $\mathbf C$ of forward propagating the image over the given convolution kernel.
  
  Assume that the bias term of the convolution is zero.

Image:
$$A=\left[\begin{matrix} 4 && 5 && 2 && 2 && 1\\
3 && 3 && 2 && 2 && 4\\
4 && 3 && 4 && 1 && 1\\
5 && 1 && 4 && 1 && 2\\
5 && 1 && 3 && 1 && 4\\
\end{matrix}\right]$$

Kernel:
$$B=\left[\begin{matrix} 4 && 3 && 3\\
5 && 5 && 5\\
2 && 4 && 3\\
\end{matrix}\right]$$

The convolution operation consists on computing the dot product of each $3 \times 3$ *sub-matrix* of $A$ times the kernel $B$ (assuming no padding and stride = 1, as specified on **(a)**)

$$Conv_B(A) = \mathbf C$$

Where $Conv_B(A)$ corresponds to the convolution operation on matrix $A$, with Toeplitz matrix of kernel $B$

And each $3 \times 3$ *sub matrix* $A_{sub_{3\times3}(ij)}$ is given by:
$$A_{sub_{3\times3}(ij)} =\left[\begin{matrix} A_{i,j} && A_{i,j+1} && A_{i,j+2}\\
A_{i+1,j} && A_{i+1,j+1} && A_{i+1,j+2}\\
A_{i+2,j} && A_{i+2,j+1} && A_{i+2,j+2}\\
\end{matrix}\right]$$

with $i,j \in [1,3]$

Using (simplified) Python notation (right bound included):

$$A_{sub_{3\times3}(ij)} = A_{[i:i+2][j:j+2]}$$

$$Conv_B(A) =\left[\begin{matrix} \left< A_{[1:3][1:3]}, B \right> && \left<A_{[1:3][2:4]}, B \right> && \left<A_{[1:3][3:5]}, B \right>\\
\left<A_{[2:3][1:3]}, B \right> && \left<A_{[2:4][2:4]}, B \right> && \left<A_{[2:4][3:5]}, B \right>\\
\left<A_{[3:5][1:3]}, B \right> && \left<A_{[3:5][2:4]}, B \right> && \left<A_{[3:5][3:5]}, B \right>\\
\end{matrix}\right]$$

To be more clear (at the risk of losing my mind writing it):

$$A=\left[\begin{matrix} 4 && 5 && 2 && 2 && 1\\
3 && 3 && 2 && 2 && 4\\
4 && 3 && 4 && 1 && 1\\
5 && 1 && 4 && 1 && 2\\
5 && 1 && 3 && 1 && 4\\
\end{matrix}\right]$$

$$Conv_B(A) = \left[\begin{matrix} \left<\left[\begin{matrix} 4 && 5 && 2\\
3 && 3 && 2\\
4 && 3 && 4\\
\end{matrix}\right], B\right> && \left<\left[\begin{matrix} 5 && 2 && 2\\
3 && 2 && 2\\
3 && 4 && 1\\
\end{matrix}\right], B\right> && \left<\left[\begin{matrix} 2 && 2 && 1\\
2 && 2 && 4\\
4 && 1 && 1\\
\end{matrix}\right], B\right>\\
\left<\left[\begin{matrix} 3 && 3 && 2\\
4 && 3 && 4\\
5 && 1 && 4\\
\end{matrix}\right], B\right> && \left<\left[\begin{matrix} 3 && 2 && 2\\
3 && 4 && 1\\
1 && 4 && 1\\
\end{matrix}\right], B\right> && \left<\left[\begin{matrix} 2 && 2 && 4\\
4 && 1 && 1\\
4 && 1 && 2\\
\end{matrix}\right], B\right>\\
\left<\left[\begin{matrix} 4 && 3 && 4\\
5 && 1 && 4\\
5 && 1 && 3\\
\end{matrix}\right], B\right> && \left<\left[\begin{matrix} 3 && 4 && 1\\
1 && 4 && 1\\
1 && 3 && 1\\
\end{matrix}\right], B\right> && \left<\left[\begin{matrix} 4 && 1 && 1\\
4 && 1 && 2\\
3 && 1 && 4\\
\end{matrix}\right], B\right> \\
\end{matrix}\right]$$

I'm not going to compute these 9 inner products by hand, but I could use something calling "programming" to help me out:


In [140]:
import numpy as np

A = np.asarray([[4., 5., 2., 2., 1.],
                [3., 3., 2., 2., 4.],
                [4., 3., 4., 1., 1.],
                [5., 1., 4., 1., 2.],
                [5., 1., 3., 1., 4.]])

B = np.asarray([[4., 3., 3.],
                [5., 5., 5.],
                [2., 4., 3.]])

print("A = \n{}".format(A))
print("B = \n{}".format(B))

A = 
[[4. 5. 2. 2. 1.]
 [3. 3. 2. 2. 4.]
 [4. 3. 4. 1. 1.]
 [5. 1. 4. 1. 2.]
 [5. 1. 3. 1. 4.]]
B = 
[[4. 3. 3.]
 [5. 5. 5.]
 [2. 4. 3.]]


In [141]:
def innerProd(X, Y):
    """Returns inner product of two matrices"""
    if X.shape != Y.shape:
        print("Error! different shape!")
        return
    res = 0
    for i in range(X.shape[0]):
        for j in range(X.shape[1]):
            res += X[i,j]*Y[i,j]
    return res

In [142]:
C = np.zeros((3,3))
for i in range(3):
    for j in range(3):
        # Get each 3x3 sub-matrix of A
        A_sub = A[i:i+3, j:j+3]
        #print("A_sub({},{}) = \n{}".format(i+1,j+1, A_sub))
        C[i, j] = innerProd(A_sub, B)
print("C = \n{}".format(C))

C = 
[[109.  92.  72.]
 [108.  85.  74.]
 [110.  74.  79.]]


|Output $\mathbf C = $|
|--|
|$$\left[\begin{matrix} 109 && 92 && 72\\
108 && 85 && 74\\
110 && 74 && 79\\
\end{matrix}\right]$$|

Let's check the result using the PyTorch library:

In [161]:
import torch
conv = torch.nn.functional.conv2d
#conv?
#conv2d(input, weight, bias=None, stride=1, padding=0, dilation=1, groups=1) -> Tensor

In [162]:
# Convert arrays to tensors
tA = torch.from_numpy(A)
tA = tA.view(1, 1, 5, 5)
tB = torch.from_numpy(B)
tB = tB.view(1, 1, 3, 3)

In [163]:
tC = conv(tA, tB)
print("C = \n{}".format(tC.numpy()))

C = 
[[[[109.  92.  72.]
   [108.  85.  74.]
   [110.  74.  79.]]]]


Which is the same result as before :)

* **(d)** Suppose the gradient backpropagated from the layers above this layer is a $3 \times 3$ matrix of all 1s.

  Write the value of the gradient with respect to the input image backpropagated out of this layer.
  
  That is, you are given that  $$\frac{\partial E}{\partial C_{ij}} = 1$$ for some scalar error $E$ and $i,j \in \{1,2,3\}$
  
  You need to compute $$\frac{\partial E}{\partial A_{ij}}$$ for $i,j \in \{1, \dots ,5\}$
  
  The chain rule should help!

#### Solution:

$$\frac{\partial E}{\partial A_{ij}} = \frac{\partial E}{\partial C_{kq}} \frac{\partial C_{kq}}{\partial A_{ij}}$$

with

>$k,q \in \{1,2,3\}$
>
>$i,j \in \{1, \dots ,5\}$

The partial derivative calculations of $C$ w.r.t. each input value $A_{ij}$ are gonna vary depending on where on the matrix is the $A_{ij}$, because the filter is **not** gonna pass the same amount of times over each $A_{ij}$:

1. Only one time:

   $$A=\left[\begin{matrix} A_{11} && - && - && - && A_{15}\\
- && - && - && - && -\\
- && - && - && - && -\\
- && - && - && - && -\\
A_{51} && - && - && - && A_{55}\\
\end{matrix}\right]$$

   $$A=\left[\begin{matrix} A_{11} && A_{12} && A_{13} && A_{14} && A_{15}\\
A_{21} && - && - && - && A_{25}\\
A_{31} && - && - && - && A_{35}\\
A_{41} && - && - && - && A_{45}\\
A_{51} && A_{52} && A_{53} && A_{54} && A_{55}\\
\end{matrix}\right]$$

2. On the second border:

    $$A=\left[\begin{matrix} - && - && - && - && -\\
- && A_{22} && A_{23} && A_{24} && -\\
- && A_{32} && - && A_{34} && -\\
- && A_{42} && A_{43} && A_{44} && -\\
- && - && - && - && -\\
\end{matrix}\right]$$

3. In the middle: 

    $$A=\left[\begin{matrix} - && - && - && - && -\\
- && - && - && - && -\\
- && - && A_{33} && - && -\\
- && - && - && - && -\\
- && - && - && - && -\\
\end{matrix}\right]$$

![hw2-convolution-partial-derivative.jpg](./img/hw2-convolution-partial-derivative.jpg)

## 1.2. Pooling

Pooling is a technique for sub-sampling and comes in different flavors, for example max-pooling, average pooling, LP-pooling.

* **(a)** List the [torch.nn modules](https://pytorch.org/docs/stable/nn.html) for the 2D versions of these pooling techniques and read what they do.

>**MaxPool2d** (kernel_size, stride=None, padding=0
>
>https://pytorch.org/docs/stable/nn.html#maxpool2d
>
> Applies Max within *kernel_size*, moving by *stride*

>**AvgPool2d**
>
> As MaxPool2d, but average

>**FractionalMaxPool2d**
>
> as MaxPool2d, but with non stochastic step

>**LPPool2d**
>
>Applies a 2D power-average pooling over an input signal composed of several input planes.
>
> $$\left( \sum_{x \in X} x^p \right)^\frac 1 p$$
>
> At $p = \infty$ , one gets **Max Pooling**
>
> At $p = 1$, one gets **Sum Pooling** (which is proportional to average pooling)

> **AdaptiveMaxPool2d**
>
> The output is of size $H \times W$, for any input size.
>
> The number of output features is equal to the number of input planes.

> **AdaptiveAvgPool2d**



* **(b)** Denote the k-th input feature maps to a pooling module as X k ∈ R H in ×W in here
H in and W in represent the input height and width, respectively.
  Let Y k ∈ R H out ×W out denote the k-th output feature map of the module where H out and W out represent the
k
output height and width, respectively. Let S i,j
be a list of the indexes of elements in the
k
k
sub-region of X used for generating Y i,j , the (i, j)-th entry of Y k . Using this notation,
give formulas for Y i,j k from three pooling modules.

* **(c)** Write out the result of applying a max-pooling module with kernel size of 2 and stride of 1 to C from Part 1.1.

* **(d)** Show how max-pooling and average pooling can be expressed in terms of LP-pooling.